### ISSUE TITLE
Should indexing a Symbol produce Indexed?

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#14351]
This seems to come up on StackOverflow a lot. Should Indexing a Symbol create an Indexed object automatically? The pro is that it would make Indexed a lot more visible to people who think that there is no way to use indexed variables. The con is that we are [already in a mess](https://github.com/sympy/sympy/issues/3539) because we make `Symbol.__call__` create a `Function`. So I'm not sure if we should do it. 

Note for new contributors: this issue is for discussion. It is **not** ready for a patch, until we agree it is a good idea.

**Comment by normalhuman** [comment]
The direct use of `Indexed(a, k)` (to which `a[k]` would presumably amount) has the downside of creating a new IndexedBase object every time. It calls `IndexedBase(a)`, and the `__new__` method of IndexedBase is not cached (unlike for Symbol), presumably because its return value is sometimes mutable (it returns a matrix if the input is a matrix).  Also, `Indexed(a, 3).base == a` would be False because IndexedBase and Symbol are different things; this looks like a source of confusion. Overall, looks like a bad practice which should not be encouraged by making it easier. 

If the issue is that users type `a[k]`, being unaware of indexed objects, we could define `Symbol.__getitem__` to throw a more informative error. 

If the issue is parsing strings containing "a[k]" (as in Stack Overflow post), then perhaps `parse_expr` should have a method `_apply_indices` similar to the existent [`_apply_functions`](https://github.com/sympy/sympy/blob/master/sympy/parsing/sympy_parser.py#L166). This method could return `Indexed(a, k)`   like `_apply_function` does with `AppliedFunction(symbol, tok)`.  The [`_group_parentheses`](https://github.com/sympy/sympy/blob/master/sympy/parsing/sympy_parser.py#L122) method would also have to deal with grouping brackets.

**Comment by asmeurer** [comment]
+1 to making sympfiy do the right thing, regardless of how Symbol works. We already should be defining Function via the parser instead of `Symbol.__call__` (#3539), so perhaps it is good to be consistent with `__getitem__` as well. It's a bit annoying to do these at the tokenize level, so we should perhaps look into using ast transformers.

**Comment by asmeurer** [comment]
I didn't know about `_apply_functions`. So does that mean that `Symbol.__call__` is unnecessary, with respect to parsing `'f(x)'` correctly?

**Comment by asmeurer** [comment]
Also with _group_parentheses already existing, the "annoying" stuff is already implemented. So maybe we can continue to hold off on ast stuff in the parser.

**Comment by normalhuman** [comment]
No, `Symbol.__call__` is still used. I was mistaken about the role of `_group_parentheses_', it's only used for "implicit application" transformation. That one happens after the tokenizing step, which is the step where f(x) becomes `Symbol('f')(Symbol('x'))`. 

Still, there seems to be room in there for a transformation rule to replace `Symbol('f')(Symbol('x'))` with `Function('f')(Symbol('x'))`, and  do likewise for square brackets. They are essentially the same issue, fixing the output of `stringify_expr`. 
 
```
>>> stringify_expr('f(x)', {}, {}, standard_transformations)
"Symbol ('f' )(Symbol ('x' ))"
>>> stringify_expr('a[k]', {}, {}, standard_transformations)
"Symbol ('a' )[Symbol ('k' )]"
```
